src/shell/shell_cmds.cpp

```diff
@@ -2572,3 +2573,121 @@ void DOS_Shell::CMD_MOVE(char* args)
 		}
 	}
 }
+
+static std::vector<std::string> search_files(std::string_view query)
+{
+	const auto save_dta = dos.dta();
+	dos.dta(dos.tables.tempdta);
+	const auto dta = DOS_DTA(dos.dta());
+
+	bool found = DOS_FindFirst(std::string(query).c_str(),
+	                           ~(DOS_ATTR_VOLUME | DOS_ATTR_DIRECTORY));
+
+	std::vector<std::string> files = {};
+	while (found) {
+		DOS_DTA::Result result = {};
+		dta.GetResult(result);
+		files.emplace_back(std::move(result.name));
+		found = DOS_FindNext();
+	}
+
+	dos.dta(save_dta);
+	
+	return files;
+}
+
+void DOS_Shell::CMD_FOR(char* args)
```

Comment:
@johnnovak , @MeAreJeenius : I feel bad that my own buggy use of passing the compiled rendered shader as a string_view across modules (which I shouldn't have done) has kind of poisoned the use of string_view. 

I think when it's used in a self-contained manner like this, it's nothing more than a read-only std::string class wrapper around a char array.

But that said: maybe we can simplify this more at the functional level and take in `FOR(const std::string_view args)`, and just work from it directly?

As a reader, seeing that we're taking a `(char* args)` tells me that we intend to modify both the pointer *and* the char values themselves (is that really the intent?)

Reply:
Yeah, all I'm saying, let's optimise for the _correct_ things. As one of my former team leads used to say: people always optimise for _something_, either consciously or unconsiously, but not always for the _correct_ thing.

Making something that takes 1ms to execute 10x faster but also making the code harder to understand by 10% is a net-negative contribution if that something is non-performance intensive and could run for 10ms or even 100ms and it wouldn't matter. That's optimising for the wrong thing. And like _meaningful_ performance improvements, these extra complexities that don't really solve anything but make the code less readable and harder to understand also add up, making the codebase overall a bit harder to maintain and understand.

All in all, the DOS programs and config file parsing are two areas where performance does not matter the _slightest_ compared to the rest of the emulator. All "inefficiencies" there are masked by the rest of the codebase by several orders of magnitude. So I would by all means make these "glue code" type of parts as readable as possible. I actually had an "outlandish" idea a while back that it would be nice if only the performance intensive parts were in C++, then all config parsing and the rest of the glue code in Python or something (no need to panic, I won't do it... too late for that, but that would have been a nice sweet-spot solution). LIke I said, you could write these "native" DOS programs in Python and no one would complain or notice the difference 😄

So yeah, the less of that iterator noise, the less mental overhead of thinking in terms of string views, pointers, copies, etc. the better in my book. Looking at this code I'm actually thinking of introducing more simple string utility functions, e.g. `std::find` is so noisy with all those iterator arguments. Ideally, I'd like non-performance intensive string mangling code like this to read like Python pseudocode.

But I won't die for it... 😄 Overall, this is no big deal, but yeah... it could be a lot simpler I think.

Reply:
Fully agree @johnnovak, and as you say: the dosbox codebase continues to deliver 👍

Looking at the original C shell-parsing code, it was a nightmare maze of single-letter pointer arithmetic without any safeguards. ⛷️ 

I'm also not a fan of all the functions that take in .begin() and .end() iterators.. like c'mon C++ committee: give us a default argument where we can  pass in the container itself (and operate on all the content).

Reply:
> The ranges library in C++20 has overloads of many of these functions. So `std::coolthing(v.begin(), v.end())` becomes `std::ranges::coolthing(v)`. The nested namespace is for some back-compat reasons that I can't recall, but at least it's an improvement, despite the verbose namespace.

Good to know. Not quite sure why we can't migrate to C++20 yet, maybe MSVC limitation or something?

Reply:
It's debian 10 (which includes retropie): unfortunately they don't have a c++20 compiler.

But debian 10 goes EOL in 9 months, so After 0.81 we can move to c++20 (knowing that debian 10 will be toast prior to the 0.82 release).

Reply:
> Many of them mutate the args, so it can't be changed to const char* either, which is a massive pain.

Yeah, that's quite horrible (the mutation)...